app/kuma-dp/pkg/dataplane/dnsproxy/component.go

```diff
@@ -0,0 +1,192 @@
+	address          string
+	port             uint32
+	done             chan struct{}
+	dnsMap           atomic.Pointer[dnsMap]
+	metrics          *metrics
+	upstreamHostPort string
+}
+
+func NewServer(address string, port uint32) *Server {
+	s := Server{
+		address: address,
+		port:    port,
+		done:    make(chan struct{}),
+		dnsMap:  atomic.Pointer[dnsMap]{},
+		metrics: newMetrics(),
+	}
+	s.dnsMap.Store(&dnsMap{ARecords: make(map[string]*dnsEntry), AAAARecords: make(map[string]*dnsEntry)})
+	return &s
+}
+
+type dnsMap struct {
+	ARecords    map[string]*dnsEntry
+	AAAARecords map[string]*dnsEntry
+}
+
+type dnsEntry struct {
+	RCode uint16
+	RR    []dns.RR
+}
+
+var _ component.GracefulComponent = &Server{}
+
+func (s *Server) Handler(res dns.ResponseWriter, req *dns.Msg) {
+	start := time.Now()
+	defer func() {
+		s.metrics.RequestDuration.Observe(time.Since(start).Seconds())
+	}()
+	response := &dns.Msg{}
+	// In case it was never loaded
+	s.dnsMap.CompareAndSwap(nil, &dnsMap{
+		ARecords:    make(map[string]*dnsEntry),
+		AAAARecords: make(map[string]*dnsEntry),
+	})
+	var dnsEntry *dnsEntry
+	if len(req.Question) > 0 { // Apparently most DNS doesn't support multiple questions so let's just support the first one
+		switch req.Question[0].Qtype {
+		case dns.TypeA:
+			// lookup in our DNS map
+			dnsMap := s.dnsMap.Load()
+			dnsEntry = dnsMap.ARecords[req.Question[0].Name]
+		case dns.TypeAAAA:
+			dnsMap := s.dnsMap.Load()
+			dnsEntry = dnsMap.AAAARecords[req.Question[0].Name]
+		}
+		log.V(1).Info("Got request", "type", req.Question[0].Qtype, "name", req.Question[0].Name, "entry", dnsEntry)
+	}
+	if dnsEntry != nil {
+		response = new(dns.Msg)
+		response.SetRcode(req, int(dnsEntry.RCode))
+		response.Authoritative = true
+		response.Answer = append(response.Answer, dnsEntry.RR...)
+	} else {
+		proxyStart := time.Now()
+		c := new(dns.Client)
+		resp, _, err := c.Exchange(req, s.upstreamHostPort)
+		if err != nil {
+			s.metrics.UpstreamRequestFailureCount.Inc()
+			log.Error(err, "Failed to write message to upstream")
+			response = &dns.Msg{}
+			response.SetRcode(req, dns.RcodeServerFailure)
+		} else {
+			response = resp
+		}
+		s.metrics.UpstreamRequestDuration.Observe(time.Since(proxyStart).Seconds())
+	}
+	err := res.WriteMsg(response)
+	if err != nil {
+		log.Error(err, "Failed to write upstreamResponse")
+	}
+}
+
+func (s *Server) Start(stop <-chan struct{}) error {
+	defer close(s.done)
+	srv := &dns.Server{Addr: net.JoinHostPort(s.address, strconv.Itoa(int(s.port))), Handler: dns.HandlerFunc(s.Handler), Net: "udp"}
+	config, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return fmt.Errorf("failed to read DNS for /etc/resolv.conf %w", err)
+	}
+	s.upstreamHostPort = net.JoinHostPort(config.Servers[0], config.Port)
+
+	done := make(chan error, 1)
+	go func() {
+		defer close(done)
+		err := srv.ListenAndServe()
+		done <- err
+	}()
+	select {
+	case <-stop:
+		err := srv.Shutdown()
```

Comment:
Question: Do we handle drainTime? If a user closes the application, we mark the dataplane as offline, but the sidecar may still have outgoing requests. We need to ensure proper handling of DNS resolution.

Reply:
Worth double checking but AFAIK this is all well handled in: https://github.com/kumahq/kuma/blob/c9644985a213dc2e957c3c3337a151a57cc46c51/app/kuma-dp/cmd/run.go#L311-L333

@jijiechen may know more because I think he did things around this recently